streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2474,6 +2476,72 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       }
     } yield javaStream
 
+  /**
+   * Converts this stream into a `scala.collection.Iterator` wrapped in a [[ZManaged]].
+   * The returned iterator will only be valid within the scope of the ZManaged.
+   */
+  def toIterator: ZManaged[R, Nothing, Iterator[Either[E, A]]] =
+    for {
+      pull    <- this.process
+      runtime <- ZIO.runtime[R].toManaged_
+    } yield {
+      new Iterator[Either[E, A]] {
```

Comment:
This is looking really good now. I think the only thing left is the internal state machine feels a little heavy, especially when we aren't managing concurrent state. Could we just implement the iterator as something like this?

```scala
new Iterator[Either[E, A]] {
  var nextTake: Take[E, A] = null
  def unsafeTake(): Unit =
    nextTake = runtime.unsafeRun(Take.fromPull(pull))
  def hasNext: Boolean = {
    if (nextTake == null) {
      unsafeTake()
    }
    !nextTake.isEnd
  }
  def next(): Either[E, A] = {
    if (nextTake == null) {
      unsafeTake()
    }
    val take = nextTake match {
      case Take.End      => throw new NoSuchElementException("next on empty iterator")
      case Take.Fail(e)  => Left(e.failureOrCause.fold(identity, c => throw FiberFailure(c)))
      case Take.Value(a) => Right(a)
    }
    nextTake = null
    take
  }
}
```

Reply:
Indeed, so much better !